Martin-school: Add REST services for management of POLLITOS SCHOOL

Martin-Mtz opened the pull request:
Change Summary:

SchoolControlApplication organized by packages and files, taking into account CRUD endpoints for grades and typical elements in a Spring Boot application with entities such as School, Student, Course, Grade, etc.

1. SchoolControlApplication.java: It is the main class that starts the Spring Boot application. Contains the main method that runs the application.
2. controller/: Contains the REST controllers that expose the APIs. In your case, the controllers for Grade, Student, Course, and School.
3. service/: Contains the business logic of the entities, such as CRUD operations and data validation.
4. model/: Contains the entities or data models (for example, Grade, Student, Course), which are mapped to database tables with JPA.
5. config/: Contains global configurations of the beans application

This PR introduces new REST services for managing grades, students, courses in the SchoolControlApplication system, facilitating the creation, reading, updating and deletion of elements associated with students within 2 schools GerardoInstitute & ZetCollege. These services allow you to manage grades, courses and students belonging to a particular school. 

Loading default data with the 'populated' profile: Updated the project configuration to ensure that when you run the populated profile, preconfigured test data is loaded into the Grade, Student, and School entities, with example initial values, which which facilitates the testing and integration process in development environments.

Latest changes

"This shouldn't be a bean, because a bean is a reusable component."
* Remove @Component Student, School, Grade, Course

"For POST and PUT request you should use @RequestBody, because in @RequestParam you have to put the info in the URL and this is dangerous"
    *studentGrades: Since this method only receives a @PathVariable and a @RequestParam, I do not consider @RequestBody necessary, since a JSON body is not being sent.
    *Some methods were updated to @RequestBody and the previous ones were preserved to exemplify another way of doing it by @RequestParam  *instructions were added to be able to test the added points
          PUT update a student
          POST to create a new student
          PUT update a course
          POST create a course
          POST create grade

The following endpoints were implemented:
Create student
Create course
Create a grade
Get all the students
Get all the courses
Get all the grades from a specific student
Edit the course information
Edit the student information
Delete all the grades of a student
Delete all the grades of a course

Extra Endpoints con  @RequestBody 
PUT update a student
POST to create a new student
PUT update a course
POST create a course
POST create grade

Extra Endpoint 
Get all the grades

How to do tests (POSTMAN): https://www.postman.com/downloads/

- Ensure that the GET, POST, PUT and DELETE methods respond as expected.
- Validate that the data uploaded when running the POPULATED profile includes sample grades for students, courses, and schools.
- Perform tests with preloaded data for ease

Create student
POST
http://localhost:8080/school/GerardoInstitute/students?id=7&firstName=Juan&lastName=Perez&age=66
http://localhost:8080/school/ZetCollege/students?id=7&firstName=Juan&lastName=Perez&age=66

Create course
POST
http://localhost:8080/school/GerardoInstitute/courses?id=1&name=Roberto&professorName=Perez
http://localhost:8080/school/ZetCollege/courses?id=1&name=Roberto&professorName=Perez

Create a grade
POST
http://localhost:8080/school/GerardoInstitute/grade?score=90&studentId=1&courseId=2
http://localhost:8080/school/ZetCollege/grade?score=90&studentId=1&courseId=2

Get all the students
http://localhost:8080/school/GerardoInstitute/getStudents
http://localhost:8080/school/ZetCollege/getStudents

Get all the courses
http://localhost:8080/school/GerardoInstitute/getCourses
http://localhost:8080/school/ZetCollege/getCourses

Get all the grades from a specific student
http://localhost:8080/school/GerardoInstitute/getGrades
http://localhost:8080/school/ZetCollege/getGrades

Edit the course information
PUT
http://localhost:8080/school/GerardoInstitute/course?id=1&name=Roberto88&professorName=Perez88
http://localhost:8080/school/ZetCollege/course?id=1&name=Roberto88&professorName=Perez88

Edit the student information
PUT
http://localhost:8080/school/GerardoInstitute/students?id=99&firstName=Juan99&lastName=Perez99&age=6699
http://localhost:8080/school/ZetCollege/students?id=99&firstName=Juan99&lastName=Perez99&age=6699

Delete all the grades of a student
http://localhost:8080/school/GerardoInstitute/studentGradesDelete?id=1
http://localhost:8080/school/ZetCollege/studentGradesDelete?id=3

Delete all the grades of a course
http://localhost:8080/school/GerardoInstitute/courseGradesDelete?id=1
http://localhost:8080/school/ZetCollege/courseGradesDelete?id=3

Extra Endpoints con  @RequestBody 

PUT update a student
http://localhost:8080/GerardoInstitute/bodyStudent
http://localhost:8080/ZetCollege/bodyStudent
{
    "id": 1,
    "firstName": "Juan",
    "lastName": "Pérez",
    "age": 19
}


POST to create a new student
http://localhost:8080/school/GerardoInstitute/student
http://localhost:8080/school/ZetCollege/student
{
    "id": 5,
    "firstName": "María",
    "lastName": "García",
    "age": 22
}

PUT update a course
http://localhost:8080/school/GerardoInstitute/getCourses
http://localhost:8080/school/ZetCollege/getCourses
{
    "id": 1,
    "name": "Historia",
    "professorName": "Prof. Sánchez"
}

POST create a course
http://localhost:8080/school/GerardoInstitute/bodyCourse
http://localhost:8080/school/ZetCollege/bodyCourse
{
    "id": 99,
    "name": "Historia",
    "professorName": "Prof. Sánchez"
}

POST create grade
http://localhost:8080/school/GerardoInstitute/bodyGrade
http://localhost:8080/school/ZetCollege/bodyGrade
{
    "score": 95,
    "studentId": 4,
    "courseId": 2
}

Extra Endpoint 
Get all the grades
http://localhost:8080/school/GerardoInstitute/getGrades
http://localhost:8080/school/ZetCollege/getGrades

Criteria to evaluate:

Creation of beans ✔️ 
Usage of the annotation in the Controller layer ✔️ 
Usage of Spring and Spring boot annotation ✔️ 
Implementation of logs ✔️ 
Naming convention ✔️ 

Extra points:
Creation of unit test or integration ⏳ 
Correct usage of interfaces and abstract classes ⏳ 
Proper instructions in the PR (description, title, steps to test) ✔️

Martin-Mtz committed:
primer commit

primer commit

Martin-Mtz committed:
Fixed

2 endpoint

Martin-Mtz committed:
CHICK SCHOOL ADVANCE

CHICK SCHOOL ADVANCE
Note: Many issues still need to be optimized that I didn't have time for

chioooo commented on src/main/java/com/iwaconsolti/school/demo/model/Student.java:
```diff
@@ -0,0 +1,15 @@
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Component
```
This shouldn't be a bean, because a bean is a reusable component.

chioooo commented on src/main/java/com/iwaconsolti/school/demo/model/Grade.java:
```diff
@@ -0,0 +1,21 @@
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Component
```
This shouldn't be a bean, because a bean is a reusable component.

Martin-Mtz replied:
done

chioooo commented on src/main/java/com/iwaconsolti/school/demo/model/Course.java:
```diff
@@ -0,0 +1,17 @@
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Component
```
This shouldn't be a bean, because a bean is a reusable component.

Martin-Mtz replied:
done

chioooo commented on src/main/java/com/iwaconsolti/school/demo/controller/GradeController.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    @PostMapping("/{schoolName}/grade")
+    public String createStudent(
```
For POST and PUT request you should use @RequestBody, because in @RequestParam you have to put the info in the URL and this is dangerous

Martin-Mtz replied:
done

chioooo reviewed (changes requested):
I see that in most of your endpoints, you're using @RequestParam, but in POST and PUT requests, you should use @RequestBody for the data being sent. Also, student, course, and grade shouldn’t be beans, as beans are meant to be reusable components.

frankisof commented on src/main/java/com/iwaconsolti/school/demo/Config/AppConfig.java:
I consider that implementing the profiles through properties would be cleaner and more understandable in your code

Martin-Mtz replied:
I didn't completely understand, the POPULATED profile is already configured in application.properties or do you mean that the data is inserted through properties? Could you give an example?

frankisof reviewed (commented):
It's a good code

Martin-Mtz committed:
Martin-school:  comments were addressed

Latest changes

"This shouldn't be a bean, because a bean is a reusable component."

Remove @component Student, School, Grade, Course
"For POST and PUT request you should use @RequestBody, because in @RequestParam you have to put the info in the URL and this is dangerous"
*studentGrades: Since this method only receives a @PathVariable and a @RequestParam, I do not consider @RequestBody necessary, since a JSON body is not being sent.
*Some methods were updated to @RequestBody and the previous ones were preserved to exemplify another way of doing it by @RequestParam *instructions were added to be able to test the added points
PUT update a student
POST to create a new student
PUT update a course
POST create a course
POST create grade

Martin-Mtz committed:
minor changes

minor changes

fjpgtt commented on src/main/java/com/iwaconsolti/school/demo/Config/AppConfig.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @Bean
+    public Student student() {return new Student();}
```
Student Course and Grade should not be beans

fjpgtt commented on src/main/java/com/iwaconsolti/school/demo/model/Course.java:
```diff
@@ -0,0 +1,15 @@
+public class Course {
+
+    private int id;
+    private String Name;
```
Name -> name

fjpgtt commented on src/main/java/com/iwaconsolti/school/demo/model/PopulateData.java:
I think you should delete this class

fjpgtt commented on src/main/java/com/iwaconsolti/school/demo/model/School.java:
```diff
@@ -0,0 +1,29 @@
+    private final List<Course> courseList;
+    private final List<Grade> gradeList;
+
+    public School() {
```
You can delete this constructor if not used

fjpgtt commented on src/main/java/com/iwaconsolti/school/demo/controller/CourseController.java:
```diff
@@ -0,0 +1,77 @@
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequestMapping("/school")
```
You can put the /{schoolName} in this requestMapping since all the endpoints have this in common

fjpgtt commented on src/main/java/com/iwaconsolti/school/demo/controller/CourseController.java:
```diff
@@ -0,0 +1,77 @@
+        this.schoolService = schoolService;
+    }
+
+    @DeleteMapping("/{schoolName}/courseGradesDelete")
```
the id that indicate the course that will remove its grades, should be in the url. The url should be something like: DELETE .../{schoolName}/course/{courseId}/grade

fjpgtt commented on src/main/java/com/iwaconsolti/school/demo/controller/CourseController.java:
```diff
@@ -0,0 +1,77 @@
+            @PathVariable String schoolName,
+            @RequestParam int id){
+
+        School school = schoolService.getSchoolByName(schoolName); // Obtener la escuela según el nombre
```
Comment in English

fjpgtt commented on src/main/java/com/iwaconsolti/school/demo/service/CourseService.java:
```diff
@@ -0,0 +1,66 @@
+            this.zetCollege = zetCollege;
+        }
+
+    public static String deleteGradesCourse(School school, int id) {
```
why is this static?

fjpgtt commented on src/main/java/com/iwaconsolti/school/demo/controller/CourseController.java:
```diff
@@ -0,0 +1,77 @@
+            @RequestParam int id){
+
+        School school = schoolService.getSchoolByName(schoolName); // Obtener la escuela según el nombre
+        return CourseService.deleteGradesCourse(school, id);
```
you should use the courseService bean

fjpgtt commented on src/main/java/com/iwaconsolti/school/demo/service/CourseService.java:
```diff
@@ -0,0 +1,66 @@
+        }
+
+    public static String deleteGradesCourse(School school, int id) {
+        for (Course existingCourse : school.getCourseList()) {
```
You should add logs to indicate what happen

fjpgtt commented on src/main/java/com/iwaconsolti/school/demo/controller/CourseController.java:
I see that you create the endpoitn with requestparameter and with request body. You should keep only the request body and remove the other

fjpgtt commented on src/main/java/com/iwaconsolti/school/demo/controller/CourseController.java:
```diff
@@ -0,0 +1,77 @@
+        return CourseService.deleteGradesCourse(school, id);
+    }
+
+    @PutMapping("/{schoolName}/bodyCourse")
```
Usage of DTOs to receive the request

fjpgtt replied:
You should pass the id of the course you want to modify in the url

fjpgtt commented on src/main/java/com/iwaconsolti/school/demo/service/CourseService.java:
```diff
@@ -0,0 +1,66 @@
+    public String createCourse(School school, Course course) {
+            for (Course existingCourse : school.getCourseList()) {
+                if (existingCourse.getId() == course.getId()) {
+                    return "Course not added; ID already exists: " + course.toString();
```
Since you are returning this message only in this case, the resposne always return the 200, when in reality this is an invalid request

fjpgtt commented on src/main/java/com/iwaconsolti/school/demo/service/CourseService.java:
```diff
@@ -0,0 +1,66 @@
+
+    public String getCourses(String schoolName){
+        if ("GerardoInstitute".equalsIgnoreCase(schoolName)) {
+            return gerardoInstitute.getCourseList().toString();
```
We should return the object course, so it is a json response, and not a string

fjpgtt commented on src/main/java/com/iwaconsolti/school/demo/service/CourseService.java:
```diff
@@ -0,0 +1,66 @@
+            if (existingCourse.getId() == course.getId()) {
+                existingCourse.setName(course.getName());
+                existingCourse.setProfessorName(course.getProfessorName());
+                return "Course update; ID already exists: " + course.toString();
```
The part about the ID already exists doesnt make a lot of sense

fjpgtt commented on src/main/java/com/iwaconsolti/school/demo/service/GradeService.java:
```diff
@@ -0,0 +1,41 @@
+
+    public String createGrade(School school, Grade grade) {
+        for (Grade existingGrade : school.getGradeList()) {
+            if (existingGrade.getMaxScore() >= 0 && existingGrade.getStudent() == grade.getStudent() && existingGrade.getCourse() == grade.getCourse()) {
```
Why you want to do this check to see if the score is more than 0?

fjpgtt commented on src/main/java/com/iwaconsolti/school/demo/controller/StudentController.java:
```diff
@@ -0,0 +1,88 @@
+        this.gradeService = gradeService;
+    }
+
+    @DeleteMapping("/{schoolName}/studentGradesDelete")
```
Same problem in this class as the Course one

fjpgtt commented on src/main/java/com/iwaconsolti/school/demo/controller/StudentController.java:
```diff
@@ -0,0 +1,88 @@
+        return studentService.deleteGradesStudent(school, id);
+    }
+
+    @PostMapping("/{schoolName}/studentsGrades") // Since this method only receives a @PathVariable and a @RequestParam, I do not consider @RequestBody necessary, since a JSON body is not being sent.
```
Why a POST when you retrieve information (GET)

fjpgtt commented on src/main/java/com/iwaconsolti/school/demo/controller/StudentController.java:
```diff
@@ -0,0 +1,88 @@
+    @PostMapping("/{schoolName}/studentsGrades") // Since this method only receives a @PathVariable and a @RequestParam, I do not consider @RequestBody necessary, since a JSON body is not being sent.
+    public String studentGrades(
+            @PathVariable String schoolName,
+            @RequestParam int id){
```
This shoudl be a pathVariable

fjpgtt commented on src/main/java/com/iwaconsolti/school/demo/service/StudentService.java:
```diff
@@ -0,0 +1,79 @@
+        if ("GerardoInstitute".equalsIgnoreCase(schoolName)) {
+            return gerardoInstitute.getStudentsList().toString();
+        } else if ("ZetCollege".equalsIgnoreCase(schoolName)) {
+            return zetCollege.getStudentsList().toString();
```
Return the object, so it is a JSON response not a simple String

fjpgtt reviewed (changes requested):
Usage of logs is missing
